pyclstm.pyx

```diff
@@ -0,0 +1,238 @@
+    then iteratively supply a line image (:py:class:`PIL.Image.Image` or
+    :py:class:`numpy.ndarray`) and the ground truth for the line to
+    :py:meth:`train`. Once finished with training, call :py:meth:`save`
+    to persist the trained model to disk.
+
+    For prediction, two methods are available. The simplest,
+    :py:meth:`recognize` takes a line image (see above) and returns the
+    recognized text as a string. If more information about the recognized
+    text is needed, use :py:meth:`recognize_chars`, which returns a generator
+    that yields :py:class:`CharPrediction` objects that contain information
+    about each character (x-offset, confidence and recognized character).
+    """
+    cdef _clstm.CLSTMOCR *_ocr
+
+    def __cinit__(self, fname=None):
+        """ Initialize the OCR engine, optionally loading a model from disk.
+
+        :param fname:   Path to pre-trained model on disk
+        :type fname:    str/unicode/bytes
+        """
+        self._ocr = new _clstm.CLSTMOCR()
+        if fname:
+            fname = safe_str(fname)
+            self.load(fname)
+
+    cpdef load(self, fname):
+        """ Load a pre-trained model from disk.
+
+        :param fname:   Path to pre-trained model on disk
+        :type fname:    str/unicode/bytes
+        """
+        fname = safe_str(fname)
+        cdef bint rv = self._ocr.maybe_load(fname)
+        if not rv:
+            raise IOError("Could not load model from {}".format(fname))
+
+    cpdef save(self, fname):
+        """ Save the model to disk.
+
+        :param fname:   Path to store model in
+        :type fname:    str/unicode/bytes
+        """
+        fname = safe_str(fname)
+        cdef bint rv = self._ocr.maybe_save(fname)
+        if not rv:
+            raise IOError("Could not save model to {}".format(fname))
+
+    cpdef prepare_training(self, graphemes, int num_hidden=100,
+                           float learning_rate=0.0001, float momentum=0.9):
+        """ Prepare training by setting the graphemes and hyperparameters.
+
+        :param graphemes:   Iterable of graphemes that are to be recognized
+                            by the OCR model, must not have duplicates
+        :type graphemes:    iterable of str/unicode
+        :param num_hidden:  Number of hidden units in the LSTM layers, larger
+                            values require more storage/memory and take longer
+                            for training and recognition, so try to find
+                            a good performance/cost tradeoff.
+        :type num_hidden:   int
+        :param learning_rate:   Learning rate for the model training
+        :type learning_rate:    float
+        :param momentum:        Momentum for the model training
+        :type momentum:         float
+        """
+        graphemes_str = u"".join(sorted(graphemes))
+        cdef vector[int] codec
+        cdef Py_ssize_t length = len(graphemes_str.encode("UTF-16")) // 2
```

Comment:
UTF-16 is a variable length encoding, which may increase code point count and produce an incorrect codec. Just exchange the length calculation by len(graphemes_str) and everything should be fine.